pkg/git/libgit2/checkout.go

```diff
@@ -179,10 +199,17 @@ func (c *CheckoutTag) Checkout(ctx context.Context, path, url string, opts *git.
 }
 
 type CheckoutCommit struct {
-	Commit string
+	Commit       string
+	LastRevision string
 }
 
 func (c *CheckoutCommit) Checkout(ctx context.Context, path, url string, opts *git.AuthOptions) (*git.Commit, error) {
+	if strings.HasSuffix(c.LastRevision, c.Commit) {
+		return nil, git.NoChangesError{
+			Message:          "no changes since last reconcilation",
+			ObservedRevision: c.LastRevision,
+		}
+	}
```

Comment:
In `CheckoutCommit` I think we may still want to go ahead and check that commit exists upstream, even if `LastRevision` matches. That should deal with the scenario in which the commit no longer exists on the upstream repository (a.k.a. deleted with force push for example).

Reply:
yep, that makes sense but there is no way to check if a commit exists upstream, without cloning the repository (except when the commit is the `HEAD`), so I don't think there are any possible optimizations here.